### ISSUE TITLE
Fail to run convert_lieder.py

### ISSUE THREAD

**Issue by weixlu** [liebharc/homr#73]
Hi, I'm running `training/datasets/convert_lieder.py` and see a few issues

### stale download url
This script needs MuseScore ro tun, and download from [url](https://cdn.jsdelivr.net/musescore/v4.2.1/MuseScore-4.2.1.240230938-x86_64.AppImage), which seems to be inaccessible (at least in my region, I'm in Asia and have tried Singapore, Japan and China in my Cloud Server)
So why don't we use the release from github? That's [https://github.com/musescore/MuseScore/releases/download/v4.2.1/MuseScore-4.2.1.240230938-x86_64.AppImage](https://github.com/musescore/MuseScore/releases/download/v4.2.1/MuseScore-4.2.1.240230938-x86_64.AppImage), the same version.

### incompatible with headless Linux
The way we run MuseScore is `os.system(MuseScore + " --force -j job.json")`. It may work on WSL, but on my headless Linux server (deployed in Cloud and connected via remote ssh), I should add `QT_QUICK_BACKEND=software QT_QPA_PLATFORM=offscreen`. 
So I suggest first detect whether it's headless then run MuseScore via
```
mscore_env = os.environ.copy()
mscore_env['QT_QUICK_BACKEND'] = 'software'
mscore_env["QT_QPA_PLATFORM"] = "offscreen"
if subprocess.run(MuseScore + " --force -j job.json", env=mscore_env, shell=True).returncode != 0:  # noqa: S605
    ....
```
Or not change the code but mention this in `Training.md`

### save time
Running `convert_lieder.py` takes quite some time, what about making it parallel? That's easy to do, just call `subprocess.Popen(MuseScore -j job-shared.json)` several times. If that make sense, I can open a PR.
Furthermore, have you thought about uploading converted datasets to [huggingface](https://huggingface.co/datasets)? In this way, other developpers don't have to convert that again. Uploading to huggingface is quite simple, I have done that before.

**Comment by weixlu** [comment]
And after musicxml/svg files are created, I see some other errors
```
INFO: Number of measures ... does not match ...

Error while processing ... Octave change isn't supported

Error while processing ... Backup duration is too long ...

Cannot merge staff at index 1: no next staff available
```
At end I see `Processed 1460/1467 files, skipped 350 files`
Is that as expected?

----------------------
Edit: Oh I got it. I have seen the discussion in #59 , and yes, liebharc said `Processed 1460/1467 files, skipped 351 files`. So I suggest mention this in `Training.md` to avoid similar confusion

**Comment by liebharc** [comment]
Hi @weixlu , not a lot of people went through the training process yet. So things are rough, and you might need to adjust things for your setup - PRs are always appreciated.

A few days ago I started a PR to update the MuseScore version, I put it on draft: https://github.com/liebharc/homr/pull/75. Note that the PR also changes the vocabulary to the pretrained models will not work when you switch to this branch. 

The original download link works for me, and it's the one from the official MuseScore website. I assume you are located in China? That might have an impact, we can switch it to GitHub if it helps in your case.

For the headless Linux, I'm also fine to do changes there. You are the first one needing that. Regarding saving time, you will likely recognize soon that this one time set up isn't the bottleneck in training. At least it's not for me, as a training run takes about 2 days. So while it's painful right now for you, it was never priority. That said, feel free to open PRs if you want to work on it!

Finally: I'm unsure about the legal situation, and can't say if uploading the files to hugging face is an option. That one aside the conversion process itself also sometimes receives improvements and bug fixes, so the end product might change which would require to keep any mirror (like hugging face) up to date. And given that so far only two people ever worked on the transformer training, the cost/benefit so far didn't look promising.

**Comment by aicelen** [comment]
I think I had the same issues with musescore as well (I'm from Germany) using ubuntu on my own machine. I also ended up using the Musescore 4.x AppImage.

I found the error log I had back then using the Musescore AppImage:
python3 -m training.train "transformer" Indexing Lieder dataset, this can up to several hours. Starting with 2690 jobs, with the first being {'in': '/home/enno/Documents/GitHub/homr/datasets/Lieder-main/flat/lc6650166.mscx', 'out': '/home/enno/Documents/GitHub/homr/datasets/Lieder-main/flat/lc6650166.musicxml'} /lib/x86_64-linux-gnu/libjack.so.0 /lib/x86_64-linux-gnu/libnss3.so QObject::moveToThread: Current thread (0x5cba65ae92d0) is not the object's thread (0x5cba65b1a520). Cannot move to target thread (0x5cba65ae92d0) qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "/home/enno/.cache/pypoetry/virtualenvs/homr-VWrplacO-py3.10/lib/python3.10/site-packages/cv2/qt/plugins" even though it was found. This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem. Available platform plugins are: xcb, offscreen. /tmp/.mount_musescAULDtM/AppRun: line 26: 10870 Aborted (core dumped) "${APPDIR}/bin/mscore4portable" "$@" Error running MuseScore

As far as I remember reinstalling the application did not fix this problem

Sadly I can't remember how exactly I fixed it...

**Comment by weixlu** [comment]
Hi @aicelen 
Yes, that's the same issue with me. I also  see somthing like ` /tmp/.mount_musescAULDtM/AppRun: line 26: 10870 Aborted (core dumped) "${APPDIR}/bin/mscore4portable" "$@`
I think the solution is just inside your log: `Available platform plugins are: xcb, offscreen`: add env QT_QPA_PLATFORM=offscreen. This works for me

**Comment by weixlu** [comment]
@liebharc hello, so let me open 2 PRs, one for download link (#76) and one for headless Linux (#77). As for the dataset convert, yes it seems not a priority. After all, trainging spends most of the time, and I'm running it. Let's see how long it takes on my server.

**Comment by liebharc** [comment]
@weixlu , good luck with the run! Let us know how it works. As we moved to convnext we struggled with an issue that the training didn't converge (blue line in https://github.com/liebharc/homr/pull/59#issuecomment-3909623478 which shows the pitch accuracy which is reported at every epoch). A successful run should get the accuracy well above 90% within a few epochs.

**Comment by weixlu** [comment]
> [@weixlu](https://github.com/weixlu) , good luck with the run! Let us know how it works. As we moved to convnext we struggled with an issue that the training didn't converge (blue line in [#59 (comment)](https://github.com/liebharc/homr/pull/59#issuecomment-3909623478) which shows the pitch accuracy which is reported at every epoch). A successful run should get the accuracy well above 90% within a few epochs.

Oh, I have seen `eval_pitch_accuracy` in the log. So far it has run 19 epochs, and result is:

Epoch | eval_pitch_accuracy | percent %
-- | -- | --
1 | 0.20662421427945066 | 20.66%
2 | 0.28162519380640444 | 28.16%
3 | 0.7631699731143173 | 76.32%
4 | 0.8521987115022585 | 85.22%
5 | 0.8771934407143643 | 87.72%
6 | 0.8762745511434344 | 87.63%
7 | 0.8840370690314624 | 88.40%
8 | 0.8861649747282045 | 88.62%
9 | 0.8842964104535117 | 88.43%
10 | 0.8875008629166021 | 88.75%
11 | 0.8886837583236469 | 88.87%
12 | 0.889773365449379 | 88.98%
13 | 0.8906036311530618 | 89.06%
14 | 0.8908797644657473 | 89.09%
15 | 0.8899077005744692 | 88.99%
16 | 0.891613010356865 | 89.16%
17 | 0.8914087090207903 | 89.14%
18 | 0.8917883923257328 | 89.18%
19 | 0.8924124909277147 | 89.24%

should I checkout to the commit you mentioned in #59 (e1034654)?

**Comment by liebharc** [comment]
I have never seen it settle at pretty much exactly 0.9. Usually it stayed either at 0.6 to 0.7 or approached 1 (0.995 or something like that). You should already have that commit as it's on main. I have no good answer to what to try there right now. If you want to test some the I would say to use that MuseScore upgrade PR which is currently open as draft, recreate the Lieder dataset (sorry 😔) and repeat the training run. But the honest answer is that I have no clear guidance right now.

**Comment by aicelen** [comment]
@weixlu What are you trying to train? Have you verified the amount of training data? 15 to 20 minutes per epoch (assuming you started training with your comment 6 hours ago) seems really fast, what hardware are you training on?

**Comment by weixlu** [comment]
> I have never seen it settle at pretty much exactly 0.9. Usually it stayed either at 0.6 to 0.7 or approached 1 (0.995 or something like that). You should already have that commit as it's on main. I have no good answer to what to try there right now. If you want to test some the I would say to use that MuseScore upgrade PR which is currently open as draft, recreate the Lieder dataset (sorry 😔) and repeat the training run. But the honest answer is that I have no clear guidance right now.

I'm trying to find out if I run the training correnctly. But I'm not sure what's the key factor I should pay attention to, I found this in my log for epoch 20. Could you please take a look?
```
{'eval_loss': 0.9119735360145569, 'eval_loss_rhythm': 0.8861593291012928, 'eval_loss_pitch': 0.007315663790762279, 'eval_loss_lift': 0.00474671600640474, 'eval_loss_consist': 0.004070392073286249, 'eval_loss_position': 0.0014051813589128174, 'eval_loss_articulations': 0.0082946751284694, 'eval_rhythm_accuracy': 0.9976388735999762, 'eval_pitch_accuracy': 0.8922856188651294, 'eval_lift_accuracy': 0.8926615706388195, 'eval_position_accuracy': 0.8933509710377202, 'eval_articulations_accuracy': 0.892201659411948, 'eval_accuracy': 0.9136277387107187, 'eval_runtime': 160.2589, 'eval_samples_per_second': 122.271, 'eval_steps_per_second': 13.591, 'epoch': 20.0}
```
And I also run the Smoke Test for epoch 20
```
Done after 52.2 seconds, SER avg: 8%
```

**Comment by weixlu** [comment]
> [@weixlu](https://github.com/weixlu) What are you trying to train? Have you verified the amount of training data? 15 to 20 minutes per epoch (assuming you started training with your comment 6 hours ago) seems really fast, what hardware are you training on?

Oh, in fact I started training yesterday, and it takes 40 mins for the first epoch. I'm running on NVIDIA A10

**Comment by liebharc** [comment]
The SER doesn't look too bad. So the training certainly works in principal. It seems however to struggle with position related information.

@weixlu what is your expectation or time commitment here? Like there is no clear answer right now and it will require some trial and error to find out what can be done. Would you be willing to help in that by running a few training runs for about 10 epochs? Or are you tight on budget and or time and you would prefer to wait until we hopefully found a solution?

**Comment by weixlu** [comment]
No problem, feel free to run, and I'll keep the server working :)

**Comment by weixlu** [comment]
And I suppose I can add batch_size to speed up? It seems the GPU has 24G memory, and only use 10G

**Comment by liebharc** [comment]
I've opened 3 draft PRs, each might affect this issue. From my data you might even able to just train them for 5 epochs (although 10 might give more meaningful data) to see if they affect the issue. I will also email the address from your git commits with a download link of all datasets as they are on my machine. This way we can ensure that we are both training on the same data. Note that this dataset will be based on https://github.com/liebharc/homr/pull/75, training it without these changes will fail.

And yes you can increase the batch size for your GPU - we could also make this an optional parameter for the training script as the optimal batch size is HW dependent. 

What you could try:
- 1 Download the files I send you and try on #75 
- 2a If it works delete the datasets and recreate them using the script
- 2b If it doesn't work rebase #78 on top of it
- 3 If it doesn't work go back to #75 and rebase #79 on top of it (you can ignore the build failure of the PR, it is just unhappy with formatting)

**Comment by weixlu** [comment]
@liebharc Hi, I'm leaving work now, and training has run 9 epochs, seems enough
I have run 3 training, #75 as run1, #75+#78 as run2, #75+#79 as run3. The stdout/stderr in training and smoke test results are here.

[result.zip](https://github.com/user-attachments/files/27959636/result.zip)

**Comment by weixlu** [comment]
Oh, by the way, the server is only accessible when I'm in the office, unavailable if I'm at home. If you need further run, tell me and I'll start it tomorrow

**Comment by liebharc** [comment]
@weixlu thanks!

Last lines from each run:

- Run 1: {'eval_loss': 0.943022608757019, 'eval_loss_rhythm': 0.8979138201409644, 'eval_loss_pitch': 0.014938170333356033, 'eval_loss_lift': 0.009074927654046239, 'eval_loss_consist': 0.005595966983495294, 'eval_loss_position': 0.002419110738612003, 'eval_loss_articulations': 0.013065964120018272, 'eval_rhythm_accuracy': 0.9944503747971968, 'eval_pitch_accuracy': 0.9932762637000201, 'eval_lift_accuracy': 0.9927864747141498, 'eval_position_accuracy': 0.9952364071212949, 'eval_articulations_accuracy': 0.9137230776522913, 'eval_accuracy': 0.9778945195969906, 'eval_runtime': 162.4988, 'eval_samples_per_second': 120.924, 'eval_steps_per_second': 13.44, 'epoch': 8.0}
- Run 2: {'eval_loss': 0.9351326823234558, 'eval_loss_rhythm': 0.8952594383896052, 'eval_loss_pitch': 0.012866703289346533, 'eval_loss_lift': 0.007694328720075506, 'eval_loss_consist': 0.004907296985838112, 'eval_loss_position': 0.0020410756675975235, 'eval_loss_articulations': 0.012351258863670006, 'eval_rhythm_accuracy': 0.995279856144235, 'eval_pitch_accuracy': 0.9949678131613016, 'eval_lift_accuracy': 0.9942074552598499, 'eval_position_accuracy': 0.9966425755000834, 'eval_articulations_accuracy': 0.9152882299559288, 'eval_accuracy': 0.9792771860042797, 'eval_runtime': 160.3561, 'eval_samples_per_second': 122.54, 'eval_steps_per_second': 13.62, 'epoch': 9.0}
- Run 3: {'eval_loss': 0.9352337718009949, 'eval_loss_rhythm': 0.8954597856208082, 'eval_loss_pitch': 0.012794569726110316, 'eval_loss_lift': 0.007829267148514564, 'eval_loss_consist': 0.005249385607165315, 'eval_loss_position': 0.0021690896909655547, 'eval_loss_articulations': 0.011746065234360569, 'eval_rhythm_accuracy': 0.9952364071212949, 'eval_pitch_accuracy': 0.9944355626302853, 'eval_lift_accuracy': 0.994696256767926, 'eval_position_accuracy': 0.9962821461052395, 'eval_articulations_accuracy': 0.9422710606795229, 'eval_accuracy': 0.9845842866608537, 'eval_runtime': 162.408, 'eval_samples_per_second': 120.992, 'eval_steps_per_second': 13.448, 'epoch': 9.0}

All accuracy values are where one would expect them. In particular pitch is > 0.99. 

You could now try to recreate the datasets and rerun the training (point 2a) from above.

**Comment by weixlu** [comment]
that's where I'm a little confused. Now I'm using the datasets you send me, but why should I recreate the datasets?

**Comment by liebharc** [comment]
Right now we don't know if the conversation scripts work properly on your machine or not. If you would recreate the datasets and run the training again then this would validate that the conversion scripts work - and we could close this issue.

That's at least the theory. I did some training runs the past days and my impression is that the training just randomly fails to converge sometimes. Possibly due to the random initialization of weights or another factor in the training process. In that case testing with the recreated dataset will not tell us much.

So it's up to you if you want to try that. If you are happy with the training run then we could also go ahead and close this issue.

As I'm curious: What are you trying to achieve with homr right now? Are you planning to extend or improve the model?

**Comment by weixlu** [comment]
Oh, I got it. And yes, I think the conversation scripts don't work properly on my machine.
- When I use your datasets, pitch is > 0.99, very good result
- When I recreate the datasets, I see pitch accuracy only 0.67

So why conversation scripts don't work properly here? I guess, MuseScore doesn't run as expected, and I do see some errors. I will take a look. 

Also, I'm happy to improve the model, and I can just use your datasets to train.

**Comment by liebharc** [comment]
Would be nice to find out what the cause of error is, as MuseScore is widely used and possibly has one of the best engraving engines. But it might also be worth to look at what https://github.com/clquwu/Clarity-OMR-Train does. It uses the same datasets as homr but has its own pipeline and uses alternatives to MuseScore for Lieder.

**Comment by weixlu** [comment]
I have figured out why my dataset performs poorly than yours. It turns out to be Muse Score fails to convert some batch of the jobs. Although there is a retry machinism, I accidentally deleted them when doing my adaption for parallel speed up ...

In short, the code itself is totally fine, and now my dataset performs just as well as yours. Now let me close this issue.
